api/ratelimit/v1alpha1/ratelimit_types.go

```diff
@@ -0,0 +1,130 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// RateLimit allows the user to limit the number of incoming requests
+// to a predefined value based on attributes within the traffic flow.
+type RateLimit struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of HTTPRoute.
+	Spec RateLimitSpec `json:"spec"`
+}
+
+// +kubebuilder:object:root=true
+
+// RateLimitList contains a list of RateLimit resources.
+type RateLimitList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []RateLimit `json:"items"`
+}
+
+// RateLimitSpec defines the desired state of RateLimit
+type RateLimitSpec struct {
+	// Type decides the scope for the RateLimits.
+	Type RateLimitType `json:"type"`
+	// Rules are a list of RateLimit matchers and limits.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RateLimitRule `json:"rules"`
+}
+
+// RateLimitType specifies the types of RateLimiting.
+// Valid RateLimitType values are:
+//
+// * "Global"
+//
+// +kubebuilder:validation:Enum=Global
+type RateLimitType string
+
+const (
+	// In this mode, the rate limits are applied across all Envoy proxy instances.
+	RateLimitTypeGlobal RateLimitType = "Global"
+)
+
+// RateLimitRule defines the semantics for matching attributes
+// from the incoming requests, and setting limits for them.
+type RateLimitRule struct {
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Matches []RateLimitMatch `json:"matches,omitempty"`
```

Comment:
in general I get a little conflicted about adding more matching criteria when this filter will already be set on a route that has a set of matching rules, seems like it will make observability a little harder?

Reply:
the `matches` here is a critical piece of the API.
* The `matches` in the `HTTPRoute` API selects which traffic flow should be sent to the backend
* The `matches` in the `RateLimitFilter` API selects which clients (deciphered from traffic flow) should be rate limited